File: components/python/src/agent/skills/order.py
"""
Order management skills for the voice sandwich shop assistant.

Provides functions to manage customer orders including adding/removing items,
confirming orders, and canceling orders. Maintains order state across the conversation.
"""

from typing import Dict, List, Optional
from datetime import datetime

# Simple in-memory order storage (keyed by thread_id in production)
# In a real system, this would be a database or cache
_orders: Dict[str, Dict] = {}


def _get_or_create_order(order_id: str = "default") -> Dict:
    """Internal helper to get or create an order."""
    if order_id not in _orders:
        _orders[order_id] = {
            "items": [],
            "status": "building",
            "created_at": datetime.now().isoformat(),
            "total": 0.0,
        }
    return _orders[order_id]


def add_to_order(
    item: str,
    quantity: int = 1,
    customizations: Optional[str] = None,
    order_id: str = "default",
) -> str:
    """
    Add an item to the customer's sandwich order.

    Args:
        item: The name of the item to add (e.g., "Turkey Club", "BLT").
        quantity: The number of items to add (default: 1).
        customizations: Optional customizations like "no tomatoes" or "extra cheese".
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message describing what was added to the order.

    Examples:
        >>> add_to_order("Turkey Club", 1)
        "Added 1 x Turkey Club to your order."

        >>> add_to_order("BLT", 2, "no mayo, extra bacon")
        "Added 2 x BLT (no mayo, extra bacon) to your order."

        >>> add_to_order("Italian Sub", 1, "on wheat bread")
        "Added 1 x Italian Sub (on wheat bread) to your order."
    """
    if quantity <= 0:
        return "Please specify a positive quantity."

    order = _get_or_create_order(order_id)

    order_item = {
        "item": item,
        "quantity": quantity,
        "customizations": customizations,
        "added_at": datetime.now().isoformat(),
    }

    order["items"].append(order_item)

    # Build confirmation message
    item_desc = f"{quantity} x {item}"
    if customizations:
        item_desc += f" ({customizations})"

    return f"Added {item_desc} to your order."


def remove_from_order(item: str, order_id: str = "default") -> str:
    """
    Remove an item from the customer's order.

    Args:
        item: The name of the item to remove.
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message or error if the item wasn't found.

    Examples:
        >>> remove_from_order("Turkey Club")
        "Removed Turkey Club from your order."

        >>> remove_from_order("Nonexistent Item")
        "Couldn't find 'Nonexistent Item' in your order."
    """
    if order_id not in _orders:
        return "You don't have any items in your order yet."

    order = _orders[order_id]

    # Search for the item (case-insensitive)
    item_lower = item.lower().strip()
    for i, order_item in enumerate(order["items"]):
        if order_item["item"].lower().strip() == item_lower:
            removed_item = order["items"].pop(i)
            item_name = removed_item["item"]
            return f"Removed {item_name} from your order."

    return f"Couldn't find '{item}' in your order."


def view_order(order_id: str = "default") -> str:
    """
    View the current order contents.

    Args:
        order_id: Order identifier (default: "default").

    Returns:
        A formatted summary of the current order.

    Examples:
        >>> view_order()
        "Your order: 1 x Turkey Club, 2 x BLT (no mayo). Total: 3 items."
    """
    if order_id not in _orders or not _orders[order_id]["items"]:
        return "Your order is empty. What would you like to order?"

    order = _orders[order_id]
    items = []

    for order_item in order["items"]:
        item_str = f"{order_item['quantity']} x {order_item['item']}"
        if order_item.get("customizations"):
            item_str += f" ({order_item['customizations']})"
        items.append(item_str)

    total_items = sum(item["quantity"] for item in order["items"])
    items_summary = ", ".join(items)

    return f"Your order: {items_summary}. Total: {total_items} item{'s' if total_items != 1 else ''}."


def confirm_order(order_id: str = "default") -> str:
    """
    Confirm and finalize the customer's order.

    This marks the order as confirmed and ready to be sent to the kitchen.

    Args:
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message with the order summary.

    Examples:
        >>> confirm_order()
        "Order confirmed: 1 x Turkey Club, 2 x BLT. Sending to the kitchen now!"
    """
    if order_id not in _orders or not _orders[order_id]["items"]:
        return "You don't have any items in your order to confirm."

    order = _orders[order_id]

    # Build order summary
    items = []
    for order_item in order["items"]:
        item_str = f"{order_item['quantity']} x {order_item['item']}"
        if order_item.get("customizations"):
            item_str += f" ({order_item['customizations']})"
        items.append(item_str)

    order_summary = ", ".join(items)

    # Mark order as confirmed
    order["status"] = "confirmed"
    order["confirmed_at"] = datetime.now().isoformat()

    return f"Order confirmed: {order_summary}. Sending to the kitchen now! Your order will be ready in about 10-15 minutes."


def cancel_order(order_id: str = "default") -> str:
    """
    Cancel the current order.

    This clears all items from the order and marks it as cancelled.

    Args:
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message.

    Examples:
        >>> cancel_order()
        "Your order has been cancelled. Let me know if you'd like to start a new order!"
    """
    if order_id not in _orders:
        return "There's no order to cancel."

    order = _orders[order_id]

    if order["status"] == "confirmed":
        # In a real system, you might want to prevent cancellation of confirmed orders
        # or handle it differently
        del _orders[order_id]
        return "Your confirmed order has been cancelled. If you placed this order recently, please check with staff."

    del _orders[order_id]
    return "Your order has been cancelled. Let me know if you'd like to start a new order!"


def modify_item(
    item: str,
    new_customizations: str,
    order_id: str = "default",
) -> str:
    """
    Modify the customizations for an existing item in the order.

    Args:
        item: The name of the item to modify.
        new_customizations: The new customizations to apply.
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message or error if the item wasn't found.

    Examples:
        >>> modify_item("Turkey Club", "no lettuce, extra tomato")
        "Updated Turkey Club: no lettuce, extra tomato"
    """
    if order_id not in _orders or not _orders[order_id]["items"]:
        return "You don't have any items in your order yet."

    order = _orders[order_id]
    item_lower = item.lower().strip()

    for order_item in order["items"]:
        if order_item["item"].lower().strip() == item_lower:
            order_item["customizations"] = new_customizations
            return f"Updated {order_item['item']}: {new_customizations}"

    return f"Couldn't find '{item}' in your order to modify."


def clear_order(order_id: str = "default") -> str:
    """
    Clear all items from the order without canceling it.

    Useful when the customer wants to start over.

    Args:
        order_id: Order identifier (default: "default").

    Returns:
        A confirmation message.

    Examples:
        >>> clear_order()
        "Cleared all items from your order. What would you like to order?"
    """
    if order_id not in _orders:
        return "Your order is already empty."

    order = _orders[order_id]
    order["items"] = []
    order["status"] = "building"

    return "Cleared all items from your order. What would you like to order?"

File: components/typescript/src/agent/skills/order.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";

/**
 * Order management skills for the sandwich shop
 * These tools handle adding, removing, confirming, and canceling orders
 */

// In-memory order state (in production, this would be stored in a database or session)
const orderState = new Map<string, { items: Array<{ item: string; quantity: number; size?: string }> }>();

export const addToOrder = tool(
  async ({ item, quantity, size, customerId = "default" }) => {
    if (!orderState.has(customerId)) {
      orderState.set(customerId, { items: [] });
    }

    const order = orderState.get(customerId)!;
    const existingItemIndex = order.items.findIndex(
      i => i.item.toLowerCase() === item.toLowerCase() && i.size === size
    );

    if (existingItemIndex >= 0) {
      order.items[existingItemIndex].quantity += quantity;
      const total = order.items[existingItemIndex].quantity;
      return `Updated your order: now ${total} ${size ? size + " " : ""}${item}${total > 1 ? "s" : ""} in total.`;
    } else {
      order.items.push({ item, quantity, size });
      return `Added ${quantity} ${size ? size + " " : ""}${item}${quantity > 1 ? "s" : ""} to your order.`;
    }
  },
  {
    name: "add_to_order",
    description: "Add an item to the customer's sandwich order. Use this when the customer wants to order a sandwich, drink, or side.",
    schema: z.object({
      item: z.string().describe("The item to add to the order (e.g., 'Turkey Club', 'BLT', 'chips')"),
      quantity: z.number().default(1).describe("Quantity of the item"),
      size: z.string().optional().describe("Size of the item: small, medium, or large"),
      customerId: z.string().optional().describe("Customer identifier"),
    }),
  }
);

export const removeFromOrder = tool(
  async ({ item, customerId = "default" }) => {
    const order = orderState.get(customerId);

    if (!order || order.items.length === 0) {
      return "Your order is already empty.";
    }

    const itemIndex = order.items.findIndex(
      i => i.item.toLowerCase() === item.toLowerCase()
    );

    if (itemIndex >= 0) {
      const removedItem = order.items.splice(itemIndex, 1)[0];
      return `Removed ${removedItem.quantity} ${removedItem.size ? removedItem.size + " " : ""}${removedItem.item}${removedItem.quantity > 1 ? "s" : ""} from your order.`;
    }

    return `I couldn't find "${item}" in your order. Would you like me to tell you what's currently in your order?`;
  },
  {
    name: "remove_from_order",
    description: "Remove an item from the customer's order. Use this when the customer wants to take something off their order.",
    schema: z.object({
      item: z.string().describe("The item to remove from the order"),
      customerId: z.string().optional().describe("Customer identifier"),
    }),
  }
);

export const confirmOrder = tool(
  async ({ customerId = "default" }) => {
    const order = orderState.get(customerId);

    if (!order || order.items.length === 0) {
      return "There's nothing in your order yet. What would you like to order?";
    }

    const orderSummary = order.items
      .map(i => `${i.quantity} ${i.size ? i.size + " " : ""}${i.item}${i.quantity > 1 ? "s" : ""}`)
      .join(", ");

    // Calculate estimated total (basic pricing)
    const estimatedTotal = order.items.reduce((sum, i) => {
      let price = 8.99; // default medium price
      if (i.size === "small") price = 6.99;
      if (i.size === "large") price = 10.99;
      return sum + (price * i.quantity);
    }, 0);

    return `Perfect! Your order is confirmed: ${orderSummary}. Your estimated total is $${estimatedTotal.toFixed(2)}. We're sending it to the kitchen now, and it'll be ready in about 10-15 minutes. Your order number is #${Math.floor(Math.random() * 1000 + 100)}.`;
  },
  {
    name: "confirm_order",
    description: "Confirm the final order with the customer and send it to the kitchen. Use this after the customer has finished ordering and is ready to complete their purchase.",
    schema: z.object({
      customerId: z.string().optional().describe("Customer identifier"),
    }),
  }
);

export const cancelOrder = tool(
  async ({ reason, customerId = "default" }) => {
    const order = orderState.get(customerId);

    if (!order || order.items.length === 0) {
      return "There's no active order to cancel.";
    }

    orderState.delete(customerId);

    if (reason) {
      return `No problem, I've cancelled your order. Reason noted: ${reason}. Is there anything else I can help you with?`;
    }

    return "Your order has been cancelled. No worries! Let me know if you'd like to start a new order.";
  },
  {
    name: "cancel_order",
    description: "Cancel the customer's current order. Use this when the customer wants to cancel or start over.",
    schema: z.object({
      reason: z.string().optional().describe("Optional reason for cancellation"),
      customerId: z.string().optional().describe("Customer identifier"),
    }),
  }
);

export const getOrderStatus = tool(
  async ({ customerId = "default" }) => {
    const order = orderState.get(customerId);

    if (!order || order.items.length === 0) {
      return "You don't have any items in your order yet. What would you like to order?";
    }

    const orderSummary = order.items
      .map(i => `${i.quantity} ${i.size ? i.size + " " : ""}${i.item}${i.quantity > 1 ? "s" : ""}`)
      .join(", ");

    return `Here's what you have so far: ${orderSummary}. Would you like to add anything else, or are you ready to confirm?`;
  },
  {
    name: "get_order_status",
    description: "Get the current status and contents of the customer's order. Use this when the customer asks what's in their order or wants to review it.",
    schema: z.object({
      customerId: z.string().optional().describe("Customer identifier"),
    }),
  }
);

export const orderSkills = [
  addToOrder,
  removeFromOrder,
  confirmOrder,
  cancelOrder,
  getOrderStatus,
];
